src/TorchSharp/DataLoader.cs

```diff
@@ -111,17 +121,50 @@ public bool MoveNext()
                         {
                             DisposeCurrent();
                             if (!MoveNextValue()) return false;
-                            List<Dictionary<string, Tensor>> dic = new();
-                            dic.Add(dataset.GetTensor(currentVal));
-                            for (var i = 1; i < batchSize; i++) {
+
+                            var tensorIndexList = new List<long> {currentVal};
+                            for (int i = 1; i < batchSize; i++) {
                                 if (!MoveNextValue()) break;
-                                dic.Add(dataset.GetTensor(currentVal));
+                                tensorIndexList.Add(currentVal);
                             }
 
+                            if (!shuffle)
+                                tensorIndexList.Reverse(); // ConcurrentBag::Add inserts data into first of the collections. So, I'll reverse them.
+
+                            ConcurrentBag<Dictionary<string, Tensor>> dic = new();
+                            var taskedBatchCount = 0;
+                            var taskBatchLock = new object();
+
+                            //Run Async
+                            foreach(var _ in Enumerable.Range(1, num_worker - 1))
+                                ThreadPool.QueueUserWorkItem(CreateBatch);
+
+                            CreateBatch(null);
+
+                            while (dic.Count < tensorIndexList.Count) { } //Wait for tensors ready
```

Comment:
Please, implement proper exception handling for failures in `GetTensor`. As it stands right now, the exception will not be propagated to `MoveNext`.

Reply:
This is related to the `Semaphore` comment. You might want to consider using .NET async tasks instead of `QueueWorkItem` to properly wire exceptions.

Reply:
Please also add a test for exception handling.

Reply:
I've test about error handling with code
```cs
class ErrorDataset : torch.utils.data.Dataset
        {
            public override long Count { get; } = 100;
            public override Dictionary<string, torch.Tensor> GetTensor(long index)
            {
                throw new System.NotImplementedException();
            }
        }
        [Fact]
        public void T()
        {
            var d = new torch.utils.data.DataLoader(new ErrorDataset(), 2);
            foreach(var x in d) {}
        }
```
in NUnit test.

This shows error like 
```
System.NotImplementedException: The method or operation is not implemented.

System.NotImplementedException
The method or operation is not implemented.
   at TorchSharp.TestDataLoader.ErrorDataset.GetTensor(Int64 index) in /home/dayo/RiderProjects/TorchSharp/test/TorchSharpTest/TestDataLoader.cs:line 138
   at TorchSharp.torch.utils.data.DataLoader.DataLoaderEnumerator.<>c__DisplayClass10_0.<MoveNext>g__ProcessPendingBatches|1() in /home/dayo/RiderProjects/TorchSharp/src/TorchSharp/DataLoader.cs:line 157
   at TorchSharp.torch.utils.data.DataLoader.DataLoaderEnumerator.MoveNext() in /home/dayo/RiderProjects/TorchSharp/src/TorchSharp/DataLoader.cs:line 142
   at TorchSharp.TestDataLoader.T() in /home/dayo/RiderProjects/TorchSharp/test/TorchSharpTest/TestDataLoader.cs:line 145
```
I think this will be enough. Stacktraces are showing fine.